comp/core/workloadmeta/filter_test.go

```diff
@@ -0,0 +1,350 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package workloadmeta
+
+import (
+	"fmt"
+	"slices"
+	"testing"
+
+	"gotest.tools/assert"
+)
+
+func TestFilterBuilder_Build(t *testing.T) {
+	dummyEntityFilterFunc := func(entity Entity) bool {
+		return len(entity.GetID().ID) == 5
+	}
+
+	tests := []struct {
+		name           string
+		builderFunc    func() *FilterBuilder
+		expectedFilter *Filter
+	}{
+		{
+			name:        "filter with default options",
+			builderFunc: NewFilterBuilder,
+			expectedFilter: &Filter{
+				source:    SourceAll,
+				kinds:     map[Kind]GenericEntityFilterFunc{},
+				eventType: EventTypeAll,
+			},
+		},
+		{
+			name: "filter with custom values",
+			builderFunc: func() *FilterBuilder {
+				return NewFilterBuilder().
+					AddKind(KindContainer).
+					AddKindWithEntityFilter(KindKubernetesPod, dummyEntityFilterFunc).
+					SetSource(SourceRuntime).
+					SetEventType(EventTypeSet)
+			},
+			expectedFilter: &Filter{
+				source: SourceRuntime,
+				kinds: map[Kind]GenericEntityFilterFunc{
+					KindContainer:     EntityFilterFuncAcceptAll,
+					KindKubernetesPod: dummyEntityFilterFunc,
+				},
+				eventType: EventTypeSet,
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(tt *testing.T) {
+			filterBuilder := test.builderFunc()
+			filter := filterBuilder.Build()
+			assert.Equal(tt, filter.source, test.expectedFilter.source)
+			assert.Equal(tt, filter.eventType, test.expectedFilter.eventType)
+
+			assert.Equal(tt, len(filter.kinds), len(test.expectedFilter.kinds))
+			for kind, entityFilter := range test.expectedFilter.kinds {
+				expectedEntityFilter, found := filter.kinds[kind]
+				assert.Check(tt, found)
+				assert.Equal(tt, fmt.Sprintf("%v", entityFilter), fmt.Sprintf("%v", expectedEntityFilter))
+			}
+		})
+	}
+}
+
+func TestFilter_MatchSource(t *testing.T) {
+	tests := []struct {
+		name              string
+		filter            *Filter
+		source            Source
+		expectMatchSource bool
+	}{
+		{
+			name:              "matched due to nil filter",
+			filter:            nil,
+			source:            "foo",
+			expectMatchSource: true,
+		},
+		{
+			name:              "matched exact source",
+			filter:            &Filter{source: "foo"},
+			source:            "foo",
+			expectMatchSource: true,
+		},
+		{
+			name:              "unmatched source",
+			filter:            &Filter{source: "foo"},
+			source:            "bar",
+			expectMatchSource: false,
+		},
+		{
+			name:              "any source should match SourceAll",
+			filter:            &Filter{},
```

Comment:
Should this be `&Filter{source:SourceAll}` ?

Reply:
Right. Not a big deal, but I thought the intention was to explicitly set SourceAll given that it's mentioned in the test name.

Reply:
That's fine. 

I updated it so it is explicitly initialised